nexus/client.go

```diff
@@ -0,0 +1,629 @@
+//
+// ⚠️ If this method completes successfully, the returned response's body must be read in its entirety and closed to
+// free up the underlying connection.
+func (c *Client) ExecuteOperation(ctx context.Context, request ExecuteOperationOptions) (*http.Response, error) {
+	result, err := c.StartOperation(ctx, request.intoStartOptions())
+	if err != nil {
+		return nil, err
+	}
+	if result.Successful != nil {
+		return result.Successful, nil
+	}
+	handle := result.Pending
+	return handle.GetResult(ctx, request.intoGetResultOptions())
+}
+
+// NewHandle gets a handle to an asynchronous operation by name and ID.
+// Does not incur a trip to the server.
+// Fails if provided any empty operation or ID, or the client has no associated ServiceBaseURL.
+func (c *Client) NewHandle(operation string, operationID string) (*OperationHandle, error) {
+	var es []error
+	if c.serviceBaseURL == nil {
+		es = append(es, errEmptyServiceBaseURL)
+	}
+	if !isValidOperationName.MatchString(operation) {
+		es = append(es, errInvalidOperationName)
+	}
+	if !isValidOperationID.MatchString(operationID) {
+		es = append(es, errInvalidOperationID)
+	}
+	if len(es) > 0 {
+		return nil, errors.Join(es...)
+	}
+	return &OperationHandle{
+		client:    c,
+		Operation: operation,
+		ID:        operationID,
+	}, nil
+}
+
+func (c *Client) sendGetOperationResultRequest(ctx context.Context, httpReq *http.Request, wait time.Duration) (*http.Response, error) {
+	if wait > 0 {
+		timeout := min(wait, c.Options.GetResultMaxTimeout)
+		if deadline, set := ctx.Deadline(); set {
+			timeout = min(timeout, time.Until(deadline))
+		}
+
+		url := httpReq.URL
+		q := url.Query()
+		q.Set(queryWait, fmt.Sprintf("%dms", timeout.Milliseconds()))
+		url.RawQuery = q.Encode()
+	}
+
+	response, err := c.Options.HTTPCaller(httpReq)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode == http.StatusOK || response.StatusCode == http.StatusNoContent {
+		return response, nil
+	}
+
+	// Do this once here and make sure it doesn't leak.
+	body, err := readAndReplaceBody(response)
+	if err != nil {
+		return nil, err
+	}
+
+	switch response.StatusCode {
+	case statusOperationRunning:
+		return nil, ErrOperationStillRunning
+	case statusOperationFailed:
+		state, err := getUnsuccessfulStateFromHeader(response, body)
+		if err != nil {
+			return nil, err
+		}
+		failure, err := failureFromResponse(response, body)
+		if err != nil {
+			return nil, err
+		}
+		return nil, &UnsuccessfulOperationError{
+			State:   state,
+			Failure: failure,
+		}
+	default:
+		return nil, newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response, body)
+	}
+}
+
+// DeliverCompletion delivers the result of a completed asynchronous operation to the provided URL.
+// If completion is an [OperationCompletionSuccessful] its body will be automatically closed.
+func (c *Client) DeliverCompletion(ctx context.Context, url string, completion OperationCompletion) error {
```

Comment:
> And therefore you can require ServiceBaseURL and get rid of the one-liner for joinURL

This would be a nice benefit here.

The other option would be to have a completion client, which I think is a bit nicer to use.

I'm changing to add a CompletionClient, it's symmetric with the CompletionHandler.

Reply:
Please don't add a new client. This doesn't need a client, their language has a client just fine. Just help them build the request.